File: examples/migrate.java
import io.github.libpasta.pasta;
import io.github.libpasta.HashUpdate;

class User {
    String pw_hash;

    public boolean auth_user() {
        String pw = pasta.read_password("Enter password:");

        HashUpdate res = pasta.verify_password_update_hash(pw_hash, pw);
        switch (res.getTag()) {
            case Updated:
                pw_hash = res.getUpdated();
            case Ok:
                return true;
        }
        return false;
    }

    public static void migrate_users(User[] users) {
        for(User user : users) {
            HashUpdate res = pasta.migrate_hash(user.pw_hash);
            switch (res.getTag()) {
                case Updated:
                    user.pw_hash = res.getUpdated();
                case Ok:
                    System.out.println("Password correct"); break;
                case Failed:
                    System.out.println("Password incorrect");
            }
        }
    }

}

File: examples/migrate.rs
extern crate libpasta;
use libpasta::rpassword::*;
use libpasta::{migrate_hash, HashUpdate};

struct User {
    // ...
    password_hash: String,
}

fn migrate_users(users: Vec<&mut User>) {
    // Step 1: Wrap old hash
    for user in users {
        let hash = user.password_hash;
        if let HashUpdate::Verified(Some(new_hash)) = migrate_hash(hash) {
            user.password_hash = new_hash;
        }
    }
}

fn auth_user(user: &mut User) {
    // Step 2: Update algorithm during log in
    let password = prompt_password_stdout("Enter password:").unwrap();

    match libpasta::verify_password_update_hash(&user.password_hash, password) {
        HashUpdate::Verified(res) => {
            if let Some(new_hash) = res {
                user.password_hash = new_hash;
            }
            println!("Password correct, new hash: \n{}", user.password_hash);
        },
        HashUpdate::Failed => {
            println!("Password incorrect, hash unchanged: \n{}", user.password_hash);
        }
    }
}